crates/storage/src/connection/pruning.rs

```diff
@@ -0,0 +1,347 @@
+//! - `canonical_blocks`
+//! - `block_signatures`
+//! - `class_definitions`
+//! - `contract_updates` (a row can be pruned if there is another row with the
+//!   same `contract_address` and a higher `block_number`)
+//! - `nonce_updates` (a row can be pruned if there is another row with the same
+//!   `contract_address_id` and a higher `block_number`)
+//! - `storage_updates` (a row can be pruned if there is another row with the
+//!   same `contract_address_id`, same `storage_address_id` and a higher
+//!   `block_number`)
+//!
+//! The number of blocks kept in the database is configurable but once chosen,
+//! it cannot change until a new database is created. It is also forbidden to
+//! enable pruning on a database that was created with it disabled (and vice
+//! versa). These restrictions are there to make it possible to prune block and
+//! (contract/nonce/storage) update data without breaking reorg handling.
+
+use anyhow::Context;
+use pathfinder_common::BlockNumber;
+
+use super::Transaction;
+use crate::prelude::{named_params, params, RowExt};
+use crate::BlockId;
+
+#[derive(Debug, Clone, Copy)]
+pub enum BlockchainHistoryMode {
+    /// Keep the entire blockchain history.
+    Archive,
+    /// Prune the blockchain history. Only keep the last `num_blocks_kept`
+    /// blocks as well as the latest block.
+    Prune { num_blocks_kept: u64 },
+}
+
+impl Transaction<'_> {
+    /// Performs pruning of the blockchain history. Last kept block's number is
+    /// `latest` - `num_blocks_kept`.
+    pub fn prune_blockchain(
+        &self,
+        latest: BlockNumber,
+        num_blocks_kept: u64,
+    ) -> anyhow::Result<()> {
+        let block_to_prune = latest
+            .checked_sub(num_blocks_kept + 1)
+            .expect("Blockchain pruning should not start before there are blocks to prune");
+
+        tracing::info!(%block_to_prune, "Running blockchain pruning");
+        let start = std::time::Instant::now();
+        self.prune_transaction_data(block_to_prune)
+            .context("Pruning transaction data")?;
+        self.prune_block_data(block_to_prune)
+            .context("Pruning block data")?;
+        tracing::debug!(pruned_block=%block_to_prune, elapsed=?start.elapsed(), "Blockchain pruning done");
+
+        Ok(())
+    }
+
+    fn prune_transaction_data(&self, block: BlockNumber) -> anyhow::Result<()> {
+        let mut transaction_stmt = self.inner().prepare_cached(
+            r"
+            DELETE FROM transactions
+            WHERE block_number = ?
+            ",
+        )?;
+        let mut transaction_hashes_stmt = self.inner().prepare_cached(
+            r"
+            DELETE FROM transaction_hashes
+            WHERE block_number = ?
+            ",
+        )?;
+        transaction_stmt.execute(params![&block])?;
+        transaction_hashes_stmt.execute(params![&block])?;
+
+        Ok(())
+    }
+
+    fn prune_block_data(&self, block: BlockNumber) -> anyhow::Result<()> {
```

Comment:
Do I understand it correctly that the algorithm is the following?

* handle trivial cases (no state updates in block)
* determine state updates in the _first_ block which is kept (`block + 1`)
* iterate over all _previous_ state updates for all updated objects (contract, nonce, storage cells) and determine which past updates are now shadowed by the updates in the first block that's kept
* remove these old updates
* check if a block that had an old update removed is now prunable and prune it

I wonder how costly these checks are during syncing. Do you have data form your resync runs?